R/response_process.R

```diff
@@ -0,0 +1,190 @@
+#' handle less happy outcomes:
+#'   * Status code in the 100s (information) or 300s (redirection). These are
+#'     unexpected.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+
+#'
+#' @details
+#' A redacted version of the `resp` input is returned in the condition (auth
+#' tokens are removed). Use functions such as `rlang::last_error()` or
+#' `rlang::catch_cnd()` to capture the condition and do a more detailed forensic
+#' examination.
+#'
+#' @param resp Object of class `response` from [httr].
+#' @param raw Logical. Whether to return parsed content (`FALSE`) or in `raw`
+#'   form.
+#'
+#' @return The content of the request.
+#' @family requests and responses
+#' @export
+response_process <- function(resp, raw = FALSE) {
+  if (httr::status_code(resp) == 204) {
+    return(TRUE)
+  }
+
+  check_response(resp)
+  content <- httr::content(resp, as = "raw")
+
+  if (raw) {
+    content
+  } else {
+    jsonlite::fromJSON(rawToChar(content), simplifyVector = FALSE)
+  }
+}
+
+check_response <- function(resp) {
+  code <- httr::status_code(resp)
+  # code could be in 100s, 200s, 300s, 400s, 500s
+
+  if (code < 200 || (code >= 300 && code < 400)) {
+    message <- c(
+      httr::http_status(resp)$message,
+      "  * HTTP status codes in the 100s and 300s are unexpected."
+    )
+    stop_request_failed(resp, message = message)
+  }
+  # code could be in 200s, 400s, 500s
+
+  if (code >= 200 && code < 300) {
+    check_for_json(resp)
+    return()
+  }
+  # code could be: 400-499, 500-599
+
+  google_error(resp)
+}
+
+stop_request_failed <- function(resp, ..., message = NULL, .subclass = NULL) {
+  code <- httr::status_code(resp)
+  class <- c(.subclass, glue("http_error_{code}"), "gargle_error_request_failed")
+
+  message <- message %||% httr::http_status(resp)$message
+  if (length(message) > 1) {
+    message <- glue_collapse(message, sep = "\n")
+  }
+
+  rlang::abort(
+    message,
+    .subclass = class,
+    code = code,
+    resp = redact_response(resp),
+    ...
+  )
+}
+
+check_for_json <- function(resp) {
+  type <- httr::http_type(resp)
+  if (!grepl("^application/json", type)) {
+    stop_need_json(resp)
+  }
+  invisible(resp)
+}
+
+stop_need_json <- function(resp) {
+  type <- httr::http_type(resp)
+  message <- c(
+    httr::http_status(resp)$message,
+    glue("  * Non-JSON content type: {type}")
+  )
+  stop_request_failed(
+    resp,
+    message = message,
+    text = httr::content(resp, as = "text")
+  )
+}
+
+google_error <- function(resp) {
+  check_for_json(resp)
+
+  code    <- httr::status_code(resp)
+  type    <- httr::http_type(resp)
+  content <- httr::content(resp, as = "raw")
+  content <- jsonlite::fromJSON(rawToChar(content), simplifyVector = FALSE)
+
+  error <- content[["error"]]
```

Comment:
This gets very if-else-y. I have identified at least 3 structurally distinct error payload styles and I'm betting there are more 😬